Let us first open one of the most recent .csv file (Requests from 2023) we have for an initial Data Exploratory Analysis

In [1]:
import pandas as pd

/Users/themistoklis/opt/anaconda3/lib/python3.9/site-packages/pandas/core/computation/expressions.py:21: UserWarning: Pandas requires version '2.8.4' or newer of 'numexpr' (version '2.8.1' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED
/Users/themistoklis/opt/anaconda3/lib/python3.9/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.4' currently installed).
  from pandas.core import (


In [3]:
df = pd.read_csv('/Users/themistoklis/Desktop/BU/Animal Reports/Analyze Boston/Boston_311_service_request_2023.csv')
df.head()

/var/folders/qf/yh559xpx4dv12z0mkyw5cvr80000gn/T/ipykernel_6361/3101638802.py:1: DtypeWarning: Columns (13) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('/Users/themistoklis/Desktop/BU/Animal Reports/Analyze Boston/Boston_311_service_request_2023.csv')


case_enquiry_id              open_dt        sla_target_dt  \
0     101005184758  2023-11-30 09:17:00  2023-12-01 09:17:52   
1     101005187668  2023-12-02 17:53:40  2023-12-05 03:30:00   
2     101005184751  2023-11-30 09:12:55  2023-12-01 09:12:57   
3     101005187656  2023-12-02 17:06:00  2023-12-11 03:30:00   
4     101005184769  2023-11-30 09:23:00                  NaN   

             closed_dt  on_time case_status  \
0  2023-11-30 23:34:48   ONTIME      Closed   
1  2023-12-03 04:26:55   ONTIME      Closed   
2  2023-11-30 09:33:07   ONTIME      Closed   
3                  NaN  OVERDUE        Open   
4  2023-12-19 11:15:33   ONTIME      Closed   

                                      closure_reason  \
0  Case Closed. Closed date : Fri Dec 01 04:34:48...   
1  Case Closed. Closed date : Sun Dec 03 09:26:55...   
2  Case Closed. Closed date : Thu Nov 30 14:33:07...   
3                                                      
4  Case Closed. Closed date : 2023-12-19 16:15:33...   

                     case_title                  subject               reason  \
0  Requests for Street Cleaning  Public Works Department      Street Cleaning   
1  Requests for Street Cleaning  Public Works Department      Street Cleaning   
2                 CE Collection  Public Works Department      Street Cleaning   
3                  PWD Graffiti  Public Works Department  Highway Maintenance   
4               Mattress Pickup  Public Works Department           Sanitation   

   ...                                  neighborhood  \
0  ...                              Greater Mattapan   
1  ...                                    Dorchester   
2  ...                                    Dorchester   
3  ...  Fenway / Kenmore / Audubon Circle / Longwood   
4  ...                                     Hyde Park   

  neighborhood_services_district     ward precinct  \
0                              9       14     1409   
1                              8  Ward 15     1507   
2                              8  Ward 15     1505   
3                             14        4     0407   
4                             10  Ward 18     1818   

                               location_street_name location_zipcode  \
0  INTERSECTION Blue Hill Ave & American Legion Hwy              NaN   
1                                     152 Homes Ave           2122.0   
2                                  108 Westville St           2124.0   
3          INTERSECTION Edgerly Rd & New Edgerly Rd              NaN   
4                                1414 Hyde Park Ave           2136.0   

    latitude  longitude                                          geom_4326  \
0  42.298179 -71.086922  0101000020E6100000A1572A2290C551C097D0CCBA2A26...   
1  42.304260 -71.064251  0101000020E6100000A96478AE1CC451C0285CF3F9F126...   
2  42.300950 -71.068161  0101000020E6100000C9C043BE5CC451C0F058B3838526...   
3  42.343791 -71.086919  0101000020E6100000E3E1211690C551C029C11E58012C...   
4  42.250370 -71.125881  0101000020E61000009D68836E0EC851C09726A51C0C20...   

                 source  
0      Constituent Call  
1  Citizens Connect App  
2       City Worker App  
3  Citizens Connect App  
4      Constituent Call  

[5 rows x 30 columns]

In [3]:
print(df.shape)

(313226, 30)


We can see that this file contains 313226 requests and they are described by 30 columns.

Lets see what are those columns corespond to.

In [4]:
print(df.columns)

Index(['case_enquiry_id', 'open_dt', 'sla_target_dt', 'closed_dt', 'on_time',
       'case_status', 'closure_reason', 'case_title', 'subject', 'reason',
       'type', 'queue', 'department', 'submitted_photo', 'closed_photo',
       'location', 'fire_district', 'pwd_district', 'city_council_district',
       'police_district', 'neighborhood', 'neighborhood_services_district',
       'ward', 'precinct', 'location_street_name', 'location_zipcode',
       'latitude', 'longitude', 'geom_4326', 'source'],
      dtype='object')


In [5]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 313226 entries, 0 to 313225
Data columns (total 30 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   case_enquiry_id                 313226 non-null  int64  
 1   open_dt                         313226 non-null  object 
 2   sla_target_dt                   247251 non-null  object 
 3   closed_dt                       283359 non-null  object 
 4   on_time                         313226 non-null  object 
 5   case_status                     313226 non-null  object 
 6   closure_reason                  313226 non-null  object 
 7   case_title                      313226 non-null  object 
 8   subject                         313226 non-null  object 
 9   reason                          313226 non-null  object 
 10  type                            313226 non-null  object 
 11  queue                           313226 non-null  object 
 12  department      

We can also have some initial info about the current file but it does not seem to be that helpful.

In [6]:
df.describe(include='all')

case_enquiry_id              open_dt        sla_target_dt  \
count      3.132260e+05               313226               247251   
unique              NaN               273520               132079   
top                 NaN  2023-08-21 12:16:31  2023-09-06 04:30:00   
freq                NaN                   37                 2137   
mean       1.010049e+11                  NaN                  NaN   
std        1.763593e+05                  NaN                  NaN   
min        1.010046e+11                  NaN                  NaN   
25%        1.010048e+11                  NaN                  NaN   
50%        1.010049e+11                  NaN                  NaN   
75%        1.010051e+11                  NaN                  NaN   
max        1.010052e+11                  NaN                  NaN   

                  closed_dt on_time case_status closure_reason  \
count                283359  313226      313226         313226   
unique               257888       2           2         277694   
top     2023-05-05 06:08:12  ONTIME      Closed                  
freq                     36  254459      283359          29789   
mean                    NaN     NaN         NaN            NaN   
std                     NaN     NaN         NaN            NaN   
min                     NaN     NaN         NaN            NaN   
25%                     NaN     NaN         NaN            NaN   
50%                     NaN     NaN         NaN            NaN   
75%                     NaN     NaN         NaN            NaN   
max                     NaN     NaN         NaN            NaN   

                 case_title                  subject  \
count                313226                   313226   
unique                 5680                       10   
top     Parking Enforcement  Public Works Department   
freq                  56942                   181343   
mean                    NaN                      NaN   
std                     NaN                      NaN   
min                     NaN                      NaN   
25%                     NaN                      NaN   
50%                     NaN                      NaN   
75%                     NaN                      NaN   
max                     NaN                      NaN   

                                  reason  ... neighborhood  \
count                             313226  ...       313010   
unique                                44  ...           24   
top     Enforcement & Abandoned Vehicles  ...   Dorchester   
freq                               61541  ...        47129   
mean                                 NaN  ...          NaN   
std                                  NaN  ...          NaN   
min                                  NaN  ...          NaN   
25%                                  NaN  ...          NaN   
50%                                  NaN  ...          NaN   
75%                                  NaN  ...          NaN   
max                                  NaN  ...          NaN   

       neighborhood_services_district     ward precinct location_street_name  \
count                          313216   313223   313103               310240   
unique                             17       55      255                82650   
top                                14  Ward 18     0306      1 City Hall Plz   
freq                            31346    17880     6174                  221   
mean                              NaN      NaN      NaN                  NaN   
std                               NaN      NaN      NaN                  NaN   
min                               NaN      NaN      NaN                  NaN   
25%                               NaN      NaN      NaN                  NaN   
50%                               NaN      NaN      NaN                  NaN   
75%                               NaN      NaN      NaN                  NaN   
max                               NaN      NaN      NaN                  NaN   

       

Just like before, due to the nature of the columns, the descripiton of the dataset does not seem really heplful.

All this information refer to every kind of request. Let us now filter out the irelevant requests.

In [24]:
#keep rows that have "animal" in the "case_title" column
df_filtered= df[df['case_title'].str.contains('animal', case=False)]
print("Out of ", df.shape[0], " requests, ", df_filtered.shape[0], " requests are related to animals")

Out of  313226  requests,  11011  requests are related to animals


In [25]:
df_filtered.head()

case_enquiry_id              open_dt        sla_target_dt  \
13      101005187223  2023-12-02 07:57:00                  NaN   
15      101005186086  2023-12-01 08:40:28  2023-12-04 08:40:29   
110     101004971366  2023-08-06 10:53:00                  NaN   
123     101004931303  2023-07-18 14:01:12  2023-07-20 04:30:00   
129     101005033448  2023-09-04 08:46:57  2023-09-06 04:30:00   

               closed_dt on_time case_status  \
13   2023-12-04 04:13:13  ONTIME      Closed   
15   2023-12-01 09:19:50  ONTIME      Closed   
110                  NaN  ONTIME        Open   
123  2023-07-18 16:11:21  ONTIME      Closed   
129  2023-09-04 09:40:59  ONTIME      Closed   

                                        closure_reason  \
13   Case Closed. Closed date : 2023-12-04 09:13:13...   
15   Case Closed. Closed date : Fri Dec 01 14:19:50...   
110                                                      
123  Case Closed. Closed date : Tue Jul 18 20:11:21...   
129  Case Closed. Closed date : Mon Sep 04 13:40:59...   

                 case_title                  subject           reason  ...  \
13   Animal Generic Request           Animal Control    Animal Issues  ...   
15      Pick up Dead Animal  Public Works Department  Street Cleaning  ...   
110  Animal Generic Request           Animal Control    Animal Issues  ...   
123     Pick up Dead Animal  Public Works Department  Street Cleaning  ...   
129     Pick up Dead Animal  Public Works Department  Street Cleaning  ...   

                               neighborhood neighborhood_services_district  \
13                                  Roxbury                             13   
15                                   Boston                              3   
110                                Brighton                             15   
123  South Boston / South Boston Waterfront                              5   
129                               South End                              6   

        ward precinct                   location_street_name location_zipcode  \
13   Ward 12     1204                         60 Edgewood St           2119.0   
15        03     0301                          102 Union Whf           2109.0   
110       21     2116                         2450 Beacon St           2135.0   
123   Ward 6     0601                               420 E St           2210.0   
129        9     0901  INTERSECTION San Juan St & Newland St              NaN   

      latitude  longitude                                          geom_4326  \
13   42.318020 -71.079171  0101000020E6100000A73E7F2111C551C0FCC959DDB428...   
15   42.365470 -71.050880  0101000020E6100000D8E0EC9F41C351C0199B74B4C72E...   
110  42.331689 -71.155636  0101000020E6100000D0F58AEFF5C951C00ED914C6742A...   
123  42.341740 -71.045460  0101000020E6100000AD3DF8D2E8C251C0D59DBE1EBE2B...   
129  42.340307 -71.074339  0101000020E610000056ED0AF9C1C451C063A30E2C8F2B...   

                   source  
13       Constituent Call  
15   Citizens Connect App  
110      Constituent Call  
123  Citizens Connect App  
129  Citizens Connect App  

[5 rows x 30 columns]

In [10]:
df_filtered.describe(include='all')

case_enquiry_id              open_dt        sla_target_dt  \
count      1.101100e+04                11011                 6849   
unique              NaN                10935                 2871   
top                 NaN  2023-01-28 07:55:57  2023-09-06 04:30:00   
freq                NaN                    4                   85   
mean       1.010049e+11                  NaN                  NaN   
std        1.658530e+05                  NaN                  NaN   
min        1.010046e+11                  NaN                  NaN   
25%        1.010048e+11                  NaN                  NaN   
50%        1.010049e+11                  NaN                  NaN   
75%        1.010051e+11                  NaN                  NaN   
max        1.010052e+11                  NaN                  NaN   

                  closed_dt on_time case_status closure_reason  \
count                 10185   11011       11011          11011   
unique                10185       2           2          10153   
top     2023-12-04 04:13:13  ONTIME      Closed                  
freq                      1   10507       10185            824   
mean                    NaN     NaN         NaN            NaN   
std                     NaN     NaN         NaN            NaN   
min                     NaN     NaN         NaN            NaN   
25%                     NaN     NaN         NaN            NaN   
50%                     NaN     NaN         NaN            NaN   
75%                     NaN     NaN         NaN            NaN   
max                     NaN     NaN         NaN            NaN   

                 case_title                  subject           reason  ...  \
count                 11011                    11011            11011  ...   
unique                    7                        3                5  ...   
top     Pick up Dead Animal  Public Works Department  Street Cleaning  ...   
freq                   6858                     6835             6828  ...   
mean                    NaN                      NaN              NaN  ...   
std                     NaN                      NaN              NaN  ...   
min                     NaN                      NaN              NaN  ...   
25%                     NaN                      NaN              NaN  ...   
50%                     NaN                      NaN              NaN  ...   
75%                     NaN                      NaN              NaN  ...   
max                     NaN                      NaN              NaN  ...   

       neighborhood neighborhood_services_district     ward precinct  \
count         11001                          11011    11011    11011   
unique           24                             17       55      255   
top      Dorchester                             10  Ward 18     0306   
freq           1896                           1153      679      246   
mean            NaN                            NaN      NaN      NaN   
std             NaN                            NaN      NaN      NaN   
min             NaN                            NaN      NaN      NaN   
25%             NaN                            NaN      NaN      NaN   
50%             NaN                            NaN      NaN      NaN   
75%             NaN                            NaN      NaN      NaN   
max             NaN                            NaN      NaN      NaN   

        location_street_name location_zipcode      latitude     longitude  \
count                  11007      9141.000000  11005.000000  11005.000000   
unique                  8584              NaN           NaN           NaN   
top     350 S Huntington Ave              NaN           NaN           NaN   
freq                      44              NaN           NaN           NaN   
mean                     NaN      2128.326660     42.323475    -71.088714   
std                      NaN        21.434365      0.033551      0.036839   
min                      NaN      2108.000000

With this initial insight, we can move onwards and apply a similar workflow for every year. We will first create a .csv file that contains only animal related requests and then perform our analysis.